magento/magento2
# Instantly apply configuration changes in the cron schedule

### Description

This pull request:
1. fixes a number of errors in the unit tests that came with the cron process
2. deletes cron jobs that have been disabled in the config, so already scheduled jobs will not execute although they have been deleted
3. deletes scheduled cron jobs that do not match their cron expression anymore, because it has been altered in the configuration.
4. changes the order of execution from running, generating, cleaning to cleaning, generating, running so that clean jobs will not run and generated jobs will run immediately.

This fixes #3380 (moved to forums)

### Fixed Issues (if relevant)

1. magento/magento2#3380: Remove scheduled jobs after changing cron settings

### Manual testing scenarios

1. Truncate the `cron_schedule` table
1. Run the `bin/magento cron:run` command
2. Look inside the `cron_schedule` table to see that `catalog_product_outdated_price_values_cleanup` is scheduled every minute
3. Open `app/code/Magento/Catalog/etc/crontab.xml` and change the cron expression from `* * * * *` to `*/10 * * * *`
4. Flush cache
5. Run the `bin/magento cron:run` command
6. See that all pending entries not matching the new expression have been removed, and nothing else
7. Change `catalog_product_outdated_price_values_cleanup` back to `* * * * *` and truncate the `cron_schedule` table
8. Run the `bin/magento cron:run` command
9. Look inside the `cron_schedule` table to see that `catalog_product_outdated_price_values_cleanup` is scheduled every minute
10. Open `app/code/Magento/Catalog/etc/crontab.xml` and remove the cron expression tag
11. Flush cache
12. Run the `bin/magento cron:run` command
13. See that all pending entries of `catalog_product_outdated_price_values_cleanup` have been removed, and nothing else

You can also play around with enabling/disabling the sitemap generator or another cron job that has configuration for enable/disable and schedules.

### Contribution checklist
